pkg/logcli/client/file.go

```diff
@@ -0,0 +1,264 @@
+	if err != nil {
+		return nil, err
+	}
+
+	return &loghttp.QueryResponse{
+		Status: "success",
+		Data: loghttp.QueryResponseData{
+			ResultType: value.Type(),
+			Result:     value,
+			Statistics: result.Statistics,
+		},
+	}, nil
+}
+
+func (f *FileClient) ListLabelNames(quiet bool, start, end time.Time) (*loghttp.LabelResponse, error) {
+	return &loghttp.LabelResponse{
+		Status: loghttp.QueryStatusSuccess,
+		Data:   f.labels,
+	}, nil
+}
+
+func (f *FileClient) ListLabelValues(name string, quiet bool, start, end time.Time) (*loghttp.LabelResponse, error) {
+	i := sort.SearchStrings(f.labels, name)
+	if i < 0 {
+		return &loghttp.LabelResponse{}, nil
+	}
+
+	return &loghttp.LabelResponse{
+		Status: loghttp.QueryStatusSuccess,
+		Data:   []string{f.labelValues[i]},
+	}, nil
+}
+
+func (f *FileClient) Series(matchers []string, start, end time.Time, quiet bool) (*loghttp.SeriesResponse, error) {
+	m := len(f.labels)
+	if m > len(f.labelValues) {
+		m = len(f.labelValues)
+	}
+
+	lbs := make(loghttp.LabelSet)
+	for i := 0; i < m; i++ {
+		lbs[f.labels[i]] = f.labelValues[i]
+	}
+
+	return &loghttp.SeriesResponse{
+		Status: loghttp.QueryStatusSuccess,
+		Data:   []loghttp.LabelSet{lbs},
+	}, nil
+}
+
+func (f *FileClient) LiveTailQueryConn(queryStr string, delayFor time.Duration, limit int, start time.Time, quiet bool) (*websocket.Conn, error) {
+	return nil, fmt.Errorf("LiveTailQuery: %w", ErrNotSupported)
+}
+
+func (f *FileClient) GetOrgID() string {
+	return f.orgID
+}
+
+type limiter struct {
+	n int
+}
+
+func (l *limiter) MaxQuerySeries(userID string) int {
+	return l.n
+}
+
+type querier struct {
+	r      io.Reader
+	labels labels.Labels
+}
+
+func (q *querier) SelectLogs(ctx context.Context, params logql.SelectLogParams) (iter.EntryIterator, error) {
+	expr, err := params.LogSelector()
+	if err != nil {
+		return nil, fmt.Errorf("failed to extract selector for logs: %w", err)
+	}
+	pipeline, err := expr.Pipeline()
+	if err != nil {
+		return nil, fmt.Errorf("failed to extract pipeline for logs: %w", err)
+	}
+	return newFileIterator(ctx, q.r, q.labels, params, pipeline.ForStream(q.labels))
+}
+
+func (q *querier) SelectSamples(ctx context.Context, params logql.SelectSampleParams) (iter.SampleIterator, error) {
+	return nil, fmt.Errorf("Metrics Query: %w", ErrNotSupported)
+}
+
+func newFileIterator(
+	ctx context.Context,
+	r io.Reader,
+	labels labels.Labels,
+	params logql.SelectLogParams,
+	pipeline logqllog.StreamPipeline,
+) (iter.EntryIterator, error) {
+
+	lr := io.LimitReader(r, defaultMaxFileSize)
+	b, err := ioutil.ReadAll(lr)
+	if err != nil {
+		return nil, err
+	}
+	lines := strings.FieldsFunc(string(b), func(r rune) bool {
+		return r == '\n'
+	})
+
+	if len(lines) == 0 {
+		return iter.NoopIterator, nil
+	}
+
+	stream := logproto.Stream{
+		Labels: labels.String(),
+	}
+
+	// reverse all the input lines if direction == FORWARD
+	if params.Direction == logproto.FORWARD {
```

Comment:
I would prefer to just range forward or backward.

Reply:
Does `FORWARD` mean all logs have to fit into memory?

Reply:
both does apparently

Reply:
Fixed!. Now we range over lines(based on direction) without explicit sorting!